File: js/src/kpke.js
/**
 * K-PKE: Encryption scheme underlying ML-KEM (FIPS 203)
 * Implements Algorithms 13, 14, and 15.
 */

import { Q, mod, fieldAdd, fieldSub, fieldMul } from './field.js';
import { ntt, nttInverse, multiplyNTTs } from './ntt.js';
import { byteEncode, byteDecode } from './encode.js';
import { compressPoly, decompressPoly } from './compress.js';
import { G, XOF, PRF } from './hash.js';
import { sampleNTT, samplePolyCBD } from './sampling.js';

/**
 * Add two polynomials coefficient-wise mod Q.
 * @param {number[]} a
 * @param {number[]} b
 * @returns {number[]}
 */
function polyAdd(a, b) {
  const c = new Array(256);
  for (let i = 0; i < 256; i++) {
    c[i] = (a[i] + b[i]) % Q;
  }
  return c;
}

/**
 * Subtract two polynomials coefficient-wise mod Q.
 * @param {number[]} a
 * @param {number[]} b
 * @returns {number[]}
 */
function polySub(a, b) {
  const c = new Array(256);
  for (let i = 0; i < 256; i++) {
    c[i] = (a[i] - b[i] + Q) % Q;
  }
  return c;
}

/**
 * Concatenate multiple Uint8Arrays.
 */
function concatBytes(...arrays) {
  const totalLen = arrays.reduce((s, a) => s + a.length, 0);
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const a of arrays) {
    result.set(a, offset);
    offset += a.length;
  }
  return result;
}

/**
 * FIPS 203 Algorithm 13: K-PKE.KeyGen(d)
 *
 * @param {Uint8Array} d - 32-byte random seed
 * @param {object} params - ML-KEM parameter set
 * @returns {{ ekPKE: Uint8Array, dkPKE: Uint8Array }}
 */
export function kpkeKeyGen(d, params) {
  const { k, eta1 } = params;

  // Step 1: (rho, sigma) = G(d || k)
  const dWithK = new Uint8Array(d.length + 1);
  dWithK.set(d);
  dWithK[d.length] = k;
  const { rho, sigma } = G(dWithK);

  // Step 2-4: Generate matrix A^ (in NTT domain) from rho
  const AHat = [];
  for (let i = 0; i < k; i++) {
    AHat[i] = [];
    for (let j = 0; j < k; j++) {
      const xof = XOF(rho, i, j);
      AHat[i][j] = sampleNTT(xof.bytes());
    }
  }

  // Step 5-7: Generate secret vector s (in normal domain, then NTT)
  let N = 0;
  const sHat = [];
  for (let i = 0; i < k; i++) {
    const prfBytes = PRF(sigma, N, 64 * eta1);
    N++;
    const si = samplePolyCBD(prfBytes, eta1);
    sHat[i] = ntt(si);
  }

  // Step 8-10: Generate error vector e (in normal domain, then NTT)
  const eHat = [];
  for (let i = 0; i < k; i++) {
    const prfBytes = PRF(sigma, N, 64 * eta1);
    N++;
    const ei = samplePolyCBD(prfBytes, eta1);
    eHat[i] = ntt(ei);
  }

  // Step 11: t^ = A^ * s^ + e^
  const tHat = [];
  for (let i = 0; i < k; i++) {
    let acc = new Array(256).fill(0);
    for (let j = 0; j < k; j++) {
      acc = polyAdd(acc, multiplyNTTs(AHat[i][j], sHat[j]));
    }
    tHat[i] = polyAdd(acc, eHat[i]);
  }

  // Step 12: Encode ekPKE = ByteEncode12(t^[0]) || ... || ByteEncode12(t^[k-1]) || rho
  const ekParts = [];
  for (let i = 0; i < k; i++) {
    ekParts.push(byteEncode(12, tHat[i]));
  }
  ekParts.push(rho);
  const ekPKE = concatBytes(...ekParts);

  // Step 13: Encode dkPKE = ByteEncode12(s^[0]) || ... || ByteEncode12(s^[k-1])
  const dkParts = [];
  for (let i = 0; i < k; i++) {
    dkParts.push(byteEncode(12, sHat[i]));
  }
  const dkPKE = concatBytes(...dkParts);

  return { ekPKE, dkPKE };
}

/**
 * FIPS 203 Algorithm 14: K-PKE.Encrypt(ekPKE, m, r)
 *
 * @param {Uint8Array} ekPKE - encryption key from kpkeKeyGen
 * @param {Uint8Array} m - 32-byte message (shared secret to encapsulate)
 * @param {Uint8Array} r - 32-byte randomness (deterministic for ML-KEM)
 * @param {object} params - ML-KEM parameter set
 * @returns {Uint8Array} ciphertext c
 */
export function kpkeEncrypt(ekPKE, m, r, params) {
  const { k, eta1, eta2, du, dv } = params;

  // Step 1: Decode t^ from ekPKE
  const tHat = [];
  for (let i = 0; i < k; i++) {
    const slice = ekPKE.slice(i * 384, (i + 1) * 384);
    tHat[i] = byteDecode(12, slice).map(c => c % Q);
  }
  const rho = ekPKE.slice(k * 384, k * 384 + 32);

  // Step 2: Generate A^ from rho (transposed: A^[j][i] for encrypt)
  const AHat = [];
  for (let i = 0; i < k; i++) {
    AHat[i] = [];
    for (let j = 0; j < k; j++) {
      const xof = XOF(rho, i, j);
      AHat[i][j] = sampleNTT(xof.bytes());
    }
  }

  // Step 3-5: Sample r vector
  let N = 0;
  const rVecHat = [];
  for (let i = 0; i < k; i++) {
    const prfBytes = PRF(r, N, 64 * eta1);
    N++;
    const ri = samplePolyCBD(prfBytes, eta1);
    rVecHat[i] = ntt(ri);
  }

  // Step 6-8: Sample e1 vector
  const e1 = [];
  for (let i = 0; i < k; i++) {
    const prfBytes = PRF(r, N, 64 * eta2);
    N++;
    e1[i] = samplePolyCBD(prfBytes, eta2);
  }

  // Step 9: Sample e2 scalar
  const e2Bytes = PRF(r, N, 64 * eta2);
  N++;
  const e2 = samplePolyCBD(e2Bytes, eta2);

  // Step 10: u = NTT^-1(A^T * r^) + e1
  // A^T[i][j] = A^[j][i]
  const u = [];
  for (let i = 0; i < k; i++) {
    let acc = new Array(256).fill(0);
    for (let j = 0; j < k; j++) {
      // A^T[i][j] = A^[j][i]
      acc = polyAdd(acc, multiplyNTTs(AHat[j][i], rVecHat[j]));
    }
    u[i] = polyAdd(nttInverse(acc), e1[i]);
  }

  // Step 11: Decode message as polynomial
  const mu = byteDecode(1, m);

  // Step 12: v = NTT^-1(t^ . r^) + e2 + Decompress_1(mu)
  let tDotR = new Array(256).fill(0);
  for (let i = 0; i < k; i++) {
    tDotR = polyAdd(tDotR, multiplyNTTs(tHat[i], rVecHat[i]));
  }
  let v = nttInverse(tDotR);
  v = polyAdd(v, e2);
  // Decompress_1(mu): each bit b maps to round(Q/2) * b
  const decompMu = new Array(256);
  for (let i = 0; i < 256; i++) {
    decompMu[i] = mu[i] === 0 ? 0 : Math.round(Q / 2);
  }
  v = polyAdd(v, decompMu);

  // Step 13-14: Compress and encode
  const c1Parts = [];
  for (let i = 0; i < k; i++) {
    const compressed = compressPoly(du, u[i]);
    c1Parts.push(byteEncode(du, compressed));
  }
  const c1 = concatBytes(...c1Parts);

  const compressedV = compressPoly(dv, v);
  const c2 = byteEncode(dv, compressedV);

  return concatBytes(c1, c2);
}

/**
 * FIPS 203 Algorithm 15: K-PKE.Decrypt(dkPKE, c)
 *
 * @param {Uint8Array} dkPKE - decryption key from kpkeKeyGen
 * @param {Uint8Array} c - ciphertext
 * @param {object} params - ML-KEM parameter set
 * @returns {Uint8Array} 32-byte message m
 */
export function kpkeDecrypt(dkPKE, c, params) {
  const { k, du, dv } = params;

  // Step 1-2: Decode and decompress u from c
  const c1Len = k * 32 * du;
  const u = [];
  for (let i = 0; i < k; i++) {
    const slice = c.slice(i * 32 * du, (i + 1) * 32 * du);
    const decoded = byteDecode(du, slice);
    u[i] = decompressPoly(du, decoded);
  }

  // Step 3: Decode and decompress v from c
  const c2 = c.slice(c1Len);
  const decodedV = byteDecode(dv, c2);
  const v = decompressPoly(dv, decodedV);

  // Step 4: Decode s^ from dkPKE
  const sHat = [];
  for (let i = 0; i < k; i++) {
    const slice = dkPKE.slice(i * 384, (i + 1) * 384);
    sHat[i] = byteDecode(12, slice).map(c => c % Q);
  }

  // Step 5: w = v - NTT^-1(s^ . NTT(u))
  let sDotU = new Array(256).fill(0);
  for (let i = 0; i < k; i++) {
    const uHat = ntt(u[i]);
    sDotU = polyAdd(sDotU, multiplyNTTs(sHat[i], uHat));
  }
  const w = polySub(v, nttInverse(sDotU));

  // Step 6: Compress_1(w) and encode as message
  const mPoly = compressPoly(1, w);
  return byteEncode(1, mPoly);
}

File: dotnet/src/PqcStandards/MlKem/Kpke.cs
namespace PqcStandards.MlKem;

/// <summary>K-PKE (internal PKE scheme) for ML-KEM per FIPS 203.</summary>
public static class Kpke
{
    /// <summary>K-PKE.KeyGen: returns (encapsulation key ek, decryption key dk).</summary>
    public static (byte[] ek, byte[] dk) KeyGen(MlKemParams p, byte[] d)
    {
        // (rho, sigma) = G(d || k)
        byte[] gInput = new byte[d.Length + 1];
        Buffer.BlockCopy(d, 0, gInput, 0, d.Length);
        gInput[d.Length] = (byte)p.K;
        byte[] g = HashFuncs.G(gInput);
        byte[] rho = g[..32];
        byte[] sigma = g[32..];

        // Generate A hat (in NTT domain)
        int[][] Ahat = new int[p.K * p.K][];
        for (int i = 0; i < p.K; i++)
            for (int j = 0; j < p.K; j++)
                Ahat[i * p.K + j] = Sampling.SampleNtt(rho, (byte)j, (byte)i);

        // Generate s (secret) and e (error)
        int[][] s = new int[p.K][];
        int[][] e = new int[p.K][];
        byte N = 0;
        for (int i = 0; i < p.K; i++)
            s[i] = Sampling.SamplePolyCbd(sigma, N++, p.Eta1);
        for (int i = 0; i < p.K; i++)
            e[i] = Sampling.SamplePolyCbd(sigma, N++, p.Eta1);

        // NTT(s), NTT(e)
        int[][] shat = new int[p.K][];
        int[][] ehat = new int[p.K][];
        for (int i = 0; i < p.K; i++)
        {
            shat[i] = Ntt.NttForward(s[i]);
            ehat[i] = Ntt.NttForward(e[i]);
        }

        // t_hat = A_hat * s_hat + e_hat
        int[][] that = new int[p.K][];
        for (int i = 0; i < p.K; i++)
        {
            that[i] = new int[256];
            for (int j = 0; j < p.K; j++)
            {
                int[] prod = Ntt.MultiplyNtts(Ahat[i * p.K + j], shat[j]);
                for (int c = 0; c < 256; c++)
                    that[i][c] = Field.FieldAdd(that[i][c], prod[c]);
            }
            for (int c = 0; c < 256; c++)
                that[i][c] = Field.FieldAdd(that[i][c], ehat[i][c]);
        }

        // Encode ek = (ByteEncode12(t_hat) || rho)
        using var ekStream = new MemoryStream();
        for (int i = 0; i < p.K; i++)
            ekStream.Write(Encode.ByteEncode(that[i], 12));
        ekStream.Write(rho);
        byte[] ek = ekStream.ToArray();

        // Encode dk = ByteEncode12(s_hat)
        using var dkStream = new MemoryStream();
        for (int i = 0; i < p.K; i++)
            dkStream.Write(Encode.ByteEncode(shat[i], 12));
        byte[] dk = dkStream.ToArray();

        return (ek, dk);
    }

    /// <summary>K-PKE.Encrypt.</summary>
    public static byte[] Encrypt(MlKemParams p, byte[] ek, byte[] m, byte[] r)
    {
        // Parse ek
        int[][] that = new int[p.K][];
        for (int i = 0; i < p.K; i++)
            that[i] = Encode.ByteDecode(ek[(384 * i)..(384 * (i + 1))], 12);
        byte[] rho = ek[(384 * p.K)..];

        // Regenerate A_hat (transposed compared to keygen)
        int[][] Ahat = new int[p.K * p.K][];
        for (int i = 0; i < p.K; i++)
            for (int j = 0; j < p.K; j++)
                Ahat[i * p.K + j] = Sampling.SampleNtt(rho, (byte)i, (byte)j);

        // Sample r_vec, e1, e2
        int[][] rvec = new int[p.K][];
        int[][] e1 = new int[p.K][];
        byte N = 0;
        for (int i = 0; i < p.K; i++)
            rvec[i] = Sampling.SamplePolyCbd(r, N++, p.Eta1);
        for (int i = 0; i < p.K; i++)
            e1[i] = Sampling.SamplePolyCbd(r, N++, p.Eta2);
        int[] e2 = Sampling.SamplePolyCbd(r, N++, p.Eta2);

        // NTT(r_vec)
        int[][] rhat = new int[p.K][];
        for (int i = 0; i < p.K; i++)
            rhat[i] = Ntt.NttForward(rvec[i]);

        // u = NTT^-1(A^T * r_hat) + e1
        int[][] u = new int[p.K][];
        for (int i = 0; i < p.K; i++)
        {
            u[i] = new int[256];
            for (int j = 0; j < p.K; j++)
            {
                int[] prod = Ntt.MultiplyNtts(Ahat[i * p.K + j], rhat[j]);
                for (int c = 0; c < 256; c++)
                    u[i][c] = Field.FieldAdd(u[i][c], prod[c]);
            }
            u[i] = Ntt.NttInverse(u[i]);
            for (int c = 0; c < 256; c++)
                u[i][c] = Field.FieldAdd(u[i][c], e1[i][c]);
        }

        // v = NTT^-1(t_hat . r_hat) + e2 + Decompress(Decode(m, 1), 1)
        int[] v = new int[256];
        for (int j = 0; j < p.K; j++)
        {
            int[] prod = Ntt.MultiplyNtts(that[j], rhat[j]);
            for (int c = 0; c < 256; c++)
                v[c] = Field.FieldAdd(v[c], prod[c]);
        }
        v = Ntt.NttInverse(v);

        int[] mu = Compress.DecompressPoly(Encode.ByteDecode(m, 1), 1);
        for (int c = 0; c < 256; c++)
            v[c] = Field.FieldAdd(Field.FieldAdd(v[c], e2[c]), mu[c]);

        // Encode ciphertext
        using var ctStream = new MemoryStream();
        for (int i = 0; i < p.K; i++)
            ctStream.Write(Encode.ByteEncode(Compress.CompressPoly(u[i], p.Du), p.Du));
        ctStream.Write(Encode.ByteEncode(Compress.CompressPoly(v, p.Dv), p.Dv));
        return ctStream.ToArray();
    }

    /// <summary>K-PKE.Decrypt.</summary>
    public static byte[] Decrypt(MlKemParams p, byte[] dk, byte[] ct)
    {
        // Parse u from ct
        int[][] u = new int[p.K][];
        int uByteLen = 32 * p.Du;
        for (int i = 0; i < p.K; i++)
        {
            int[] compressed = Encode.ByteDecode(ct[(uByteLen * i)..(uByteLen * (i + 1))], p.Du);
            u[i] = Compress.DecompressPoly(compressed, p.Du);
        }

        // Parse v from ct
        int vOffset = uByteLen * p.K;
        int[] vCompressed = Encode.ByteDecode(ct[vOffset..], p.Dv);
        int[] v = Compress.DecompressPoly(vCompressed, p.Dv);

        // Parse s_hat from dk
        int[][] shat = new int[p.K][];
        for (int i = 0; i < p.K; i++)
            shat[i] = Encode.ByteDecode(dk[(384 * i)..(384 * (i + 1))], 12);

        // w = v - NTT^-1(s_hat . NTT(u))
        int[] w = new int[256];
        for (int j = 0; j < p.K; j++)
        {
            int[] uhat = Ntt.NttForward(u[j]);
            int[] prod = Ntt.MultiplyNtts(shat[j], uhat);
            for (int c = 0; c < 256; c++)
                w[c] = Field.FieldAdd(w[c], prod[c]);
        }
        w = Ntt.NttInverse(w);
        for (int c = 0; c < 256; c++)
            w[c] = Field.FieldSub(v[c], w[c]);

        return Encode.ByteEncode(Compress.CompressPoly(w, 1), 1);
    }
}
